--- /lib/modules/2.6.30.9-scst-debug/build/scripts/Makefile.lib.orig	2009-06-10 05:05:27.000000000 +0200
+++ /lib/modules/2.6.30.9-scst-debug/build/scripts/Makefile.lib	2009-10-27 20:20:19.000000000 +0100
@@ -111,7 +111,7 @@
 
 orig_c_flags   = $(KBUILD_CPPFLAGS) $(KBUILD_CFLAGS) $(KBUILD_SUBDIR_CCFLAGS) \
                  $(ccflags-y) $(CFLAGS_$(basetarget).o)
-_c_flags       = $(filter-out $(CFLAGS_REMOVE_$(basetarget).o), $(orig_c_flags))
+_c_flags       = $(filter-out $(CFLAGS_REMOVE_$(basetarget).o), $(orig_c_flags)) $(PRE_CFLAGS)
 _a_flags       = $(KBUILD_CPPFLAGS) $(KBUILD_AFLAGS) $(KBUILD_SUBDIR_ASFLAGS) \
                  $(asflags-y) $(AFLAGS_$(basetarget).o)
 _cpp_flags     = $(KBUILD_CPPFLAGS) $(cppflags-y) $(CPPFLAGS_$(@F))
