diff --git a/Makefile b/Makefile
index fd80c6e..09ca4ea 100644
--- a/Makefile
+++ b/Makefile
@@ -390,6 +390,7 @@ USERINCLUDE    := \
 # Use LINUXINCLUDE when you must reference the include/ directory.
 # Needed to be compatible with the O= option
 LINUXINCLUDE    := \
+		$(PRE_CFLAGS) \
 		-I$(srctree)/arch/$(hdr-arch)/include \
 		-Iarch/$(hdr-arch)/include/generated \
 		$(if $(KBUILD_SRC), -I$(srctree)/include) \
