--- linux-2.6.18/scripts/Makefile.lib	2009-05-10 20:00:05.000000000 +0200
+++ linux-2.6.18/scripts/Makefile.lib	2009-05-10 20:00:25.000000000 +0200
@@ -108,7 +108,7 @@
 __cpp_flags     =                          $(call flags,_cpp_flags)
 endif
 
-c_flags        = -Wp,-MD,$(depfile) $(NOSTDINC_FLAGS) $(CPPFLAGS) \
+c_flags        = -Wp,-MD,$(depfile) $(NOSTDINC_FLAGS) $(PRE_CFLAGS) $(CPPFLAGS) \
 		 $(__c_flags) $(modkern_cflags) \
 		 -D"KBUILD_STR(s)=\#s" $(basename_flags) $(modname_flags)
 
