diff --git a/mm/readahead.c b/mm/readahead.c
index 133b6d5..fd3df66 100644
--- a/mm/readahead.c
+++ b/mm/readahead.c
@@ -479,5 +479,8 @@ page_cache_async_readahead(struct address_space *mapping,
 
 	/* do read-ahead */
 	ondemand_readahead(mapping, ra, filp, true, offset, req_size);
+
+	if (PageUptodate(page))
+		blk_run_backing_dev(mapping->backing_dev_info, NULL);
 }
 EXPORT_SYMBOL_GPL(page_cache_async_readahead);
