Index: iscsi-scst/kernel/Makefile
===================================================================
--- iscsi-scst/kernel/Makefile	(revision 324)
+++ iscsi-scst/kernel/Makefile	(working copy)
@@ -23,9 +23,9 @@ SCST_INC_DIR := $(SUBDIRS)/../../scst/in
 EXTRA_CFLAGS += -I$(src)/../include -I$(SCST_INC_DIR)
 # -Wextra -Wno-unused-parameter
 
-EXTRA_CFLAGS += -DCONFIG_SCST_EXTRACHECKS
-#EXTRA_CFLAGS += -DCONFIG_SCST_TRACING
-EXTRA_CFLAGS += -DCONFIG_SCST_DEBUG -g
+#EXTRA_CFLAGS += -DCONFIG_SCST_EXTRACHECKS
+EXTRA_CFLAGS += -DCONFIG_SCST_TRACING
+#EXTRA_CFLAGS += -DCONFIG_SCST_DEBUG -g
 
 #EXTRA_CFLAGS += -DCONFIG_SCST_ISCSI_DEBUG_DIGEST_FAILURES
 
