Index: qla2x00t/qla2x00-target/Makefile
===================================================================
--- qla2x00t/qla2x00-target/Makefile	(revision 542)
+++ qla2x00t/qla2x00-target/Makefile	(working copy)
@@ -31,14 +31,14 @@
 SCST_INC_DIR := $(SUBDIRS)/../../scst/include
 SCST_DIR := $(shell pwd)/../../scst/src
 
-EXTRA_CFLAGS += -I$(SCST_INC_DIR)
+EXTRA_CFLAGS += -I$(SCST_INC_DIR) -Wextra -Wno-unused-parameter \
+		-Wno-missing-field-initializers
 
 INSTALL_DIR := /lib/modules/$(shell uname -r)/extra
 
-EXTRA_CFLAGS += -DCONFIG_SCST_EXTRACHECKS 
-#EXTRA_CFLAGS += -DCONFIG_SCST_TRACING
-EXTRA_CFLAGS += -DCONFIG_SCST_DEBUG -g -W -Wno-unused-parameter \
-		-Wno-missing-field-initializers
+#EXTRA_CFLAGS += -DCONFIG_SCST_EXTRACHECKS 
+EXTRA_CFLAGS += -DCONFIG_SCST_TRACING
+#EXTRA_CFLAGS += -DCONFIG_SCST_DEBUG -g -W
 #EXTRA_CFLAGS += -DCONFIG_QLA_TGT_DEBUG_WORK_IN_THREAD
 
 ifeq ($(KVER),)
