Index: qla2x00t/qla2x00-target/Makefile
===================================================================
--- qla2x00t/qla2x00-target/Makefile	(revision 213)
+++ qla2x00t/qla2x00-target/Makefile	(working copy)
@@ -35,9 +35,9 @@ EXTRA_CFLAGS += -I$(SCST_INC_DIR) -DFC_T
 
 INSTALL_DIR := /lib/modules/$(shell uname -r)/extra
 
-EXTRA_CFLAGS += -DCONFIG_SCST_EXTRACHECKS 
+#EXTRA_CFLAGS += -DCONFIG_SCST_EXTRACHECKS 
 #EXTRA_CFLAGS += -DCONFIG_SCST_TRACING
-EXTRA_CFLAGS += -DDEBUG_TGT -g
+#EXTRA_CFLAGS += -DDEBUG_TGT -g
 #EXTRA_CFLAGS += -DDEBUG_WORK_IN_THREAD
 
 ifeq ($(KVER),)
