diff -uprN -X/home/stasiu/work/linux-2.6/Documentation/dontdiff qla_isp/linux/isp_pci.c qla_isp-1.0.0/linux/isp_pci.c
--- qla_isp/linux/isp_pci.c	2008-07-08 09:38:11.000000000 +0200
+++ qla_isp-1.0.0/linux/isp_pci.c	2008-07-06 12:50:35.000000000 +0200
@@ -536,10 +536,7 @@
     if (isp_debug) {
         isp->isp_dblev = isp_debug;
     } else {
-        isp->isp_dblev = ISP_LOGCONFIG|ISP_LOGINFO|ISP_LOGWARN|ISP_LOGERR;
-#ifdef  ISP_TARGET_MODE
-        isp->isp_dblev |= ISP_LOGTINFO;
-#endif
+        isp->isp_dblev = ISP_LOGERR;
     }
 
     pci_read_config_word(pdev, PCI_COMMAND, &cmd);
diff -uprN -X/home/stasiu/work/linux-2.6/Documentation/dontdiff qla_isp/linux/isp_scst.c qla_isp-1.0.0/linux/isp_scst.c
--- qla_isp/linux/isp_scst.c	2008-07-08 09:38:11.000000000 +0200
+++ qla_isp-1.0.0/linux/isp_scst.c	2008-07-06 12:49:07.000000000 +0200
@@ -156,7 +156,7 @@ struct bus {
     struct scst_proc_data    proc_data;
 };
 
-#define DEBUG 1
+//#define DEBUG 1
 
 #ifdef DEBUG
 #define    SDprintk     if (debug) printk
