Index: scst/src/Makefile
===================================================================
--- scst/src/Makefile	(revision 233)
+++ scst/src/Makefile	(working copy)
@@ -114,7 +114,7 @@ EXTRA_CFLAGS += -I$(SCST_INC_DIR) -Wextr
 
 #EXTRA_CFLAGS += -DSTRICT_SERIALIZING
 
-EXTRA_CFLAGS += -DEXTRACHECKS
+#EXTRA_CFLAGS += -DEXTRACHECKS
 
 #EXTRA_CFLAGS += -DUSE_EXPECTED_VALUES
 #EXTRA_CFLAGS += -DALLOW_PASSTHROUGH_IO_SUBMIT_IN_SIRQ
@@ -122,9 +122,9 @@ EXTRA_CFLAGS += -DEXTRACHECKS
 
 #EXTRA_CFLAGS += -fno-inline
 
-#EXTRA_CFLAGS += -DTRACING
+EXTRA_CFLAGS += -DTRACING
 
-EXTRA_CFLAGS += -DDEBUG -g
+#EXTRA_CFLAGS += -DDEBUG -g
 #EXTRA_CFLAGS += -DDEBUG_TM -DTM_DBG_GO_OFFLINE=0
 #EXTRA_CFLAGS += -DDEBUG_RETRY
 #EXTRA_CFLAGS += -DDEBUG_OOM
Index: scst/src/dev_handlers/Makefile
===================================================================
--- scst/src/dev_handlers/Makefile	(revision 230)
+++ scst/src/dev_handlers/Makefile	(working copy)
@@ -68,10 +68,10 @@ endif
 
 EXTRA_CFLAGS += -I$(SUBDIRS) -I$(SCST_INC_DIR) -Wextra -Wno-unused-parameter
 
-EXTRA_CFLAGS += -DEXTRACHECKS
+#EXTRA_CFLAGS += -DEXTRACHECKS
 
-#EXTRA_CFLAGS += -DTRACING
-EXTRA_CFLAGS += -DDEBUG -g
+EXTRA_CFLAGS += -DTRACING
+#EXTRA_CFLAGS += -DDEBUG -g
 
 clean:
 	rm -f *.o *.ko .*.cmd *.mod.c .*.d .depend Modules.symvers Module.symvers
