Index: usr/fileio/Makefile
===================================================================
--- usr/fileio/Makefile	(revision 324)
+++ usr/fileio/Makefile	(working copy)
@@ -28,9 +28,9 @@ CFLAGS += -O2 -Wall -Wextra -Wno-unused-
 PROGS = fileio_tgt
 LIBS = -lpthread
 
-CFLAGS += -DEXTRACHECKS
+#CFLAGS += -DEXTRACHECKS
 #CFLAGS += -DTRACING
-CFLAGS += -DDEBUG -g
+#CFLAGS += -DDEBUG -g
 
 #CFLAGS += -DDEBUG_NOMEM
 #CFLAGS += -DDEBUG_SENSE
