diff -upkr linux-2.6.32.x86_64/include/linux/Kbuild linux-2.6.32.x86_64/include/linux/Kbuild
--- linux-2.6.32.x86_64/include/linux/Kbuild    2009-12-03 04:51:21.000000000 +0100
+++ linux-2.6.32.x86_64/include/linux/Kbuild    2010-08-27 13:14:53.767830262 +0200
@@ -112,6 +112,7 @@ header-y += map_to_7segment.h
 header-y += matroxfb.h
 header-y += meye.h
 header-y += minix_fs.h
+header-y += mm.h
 header-y += mmtimer.h
 header-y += mqueue.h
 header-y += mtio.h
