--- Makefile	2004-08-18 13:20:55.434402848 +0400
+++ Makefile.scsi	2004-08-13 18:01:13.000000000 +0400
@@ -81,6 +81,7 @@ obj-$(CONFIG_SCSI_QLOGIC_ISP)	+= qlogici
 obj-$(CONFIG_SCSI_QLOGIC_FC)	+= qlogicfc.o 
 obj-$(CONFIG_SCSI_QLOGIC_1280)	+= qla1280.o 
 obj-$(CONFIG_SCSI_QLA2XXX)	+= qla2xxx/
+obj-$(CONFIG_SCST)		+= scsi_tgt/
 obj-$(CONFIG_SCSI_PAS16)	+= pas16.o
 obj-$(CONFIG_SCSI_SEAGATE)	+= seagate.o
 obj-$(CONFIG_SCSI_FD_8xx)	+= seagate.o
