diff -uprN orig/linux-2.6.24/drivers/scsi/Makefile linux-2.6.24/drivers/scsi/Makefile
--- orig/linux-2.6.24/drivers/scsi/Makefile	2008-01-24 23:58:37.000000000 +0100
+++ linux-2.6.24/drivers/scsi/Makefile	2008-05-09 13:05:36.000000000 +0200
@@ -88,6 +88,7 @@ obj-$(CONFIG_PCMCIA_QLOGIC)	+= qlogicfas
 obj-$(CONFIG_SCSI_QLOGIC_1280)	+= qla1280.o 
 obj-$(CONFIG_SCSI_QLA_FC)	+= qla2xxx/
 obj-$(CONFIG_SCSI_QLA_ISCSI)	+= qla4xxx/
+obj-$(CONFIG_SCSI_TARGET)	+= scsi_tgt/
 obj-$(CONFIG_SCSI_LPFC)		+= lpfc/
 obj-$(CONFIG_SCSI_PAS16)	+= pas16.o
 obj-$(CONFIG_SCSI_SEAGATE)	+= seagate.o
